### PR TITLE
Add support for reductions to TensorIterator

### PR DESCRIPTION
**Pull request by colesbury**
This adds support for reductions like sum() and mul() to TensorIterator.
Performance is similar to existing optimized code for CPU, and generally
better than existing code for CUDA kernels.

The templatized CUDA kernel requires fewer instantiations than the
existing THCReduce/THCReduceAll code. For example, sum() previously
generated 43 CUDA kernels, while it now requires only one (larger)
CUDA kernel. I suspect this should reduce code-size and
compilation time, but I haven't measured it.

Below are timings for sum() on [CPU](https://ark.intel.com/products/81908/Intel-Xeon-Processor-E5-2680-v3-30M-Cache-2_50-GHz) (12 threads and 1 thread) and CUDA with various tensor sizes.

CPU

| Reduction (dim)      | Master  | PR      | Master (1 thread) | PR (1 thread) |
|----------------------|---------|---------|-------------------|---------------|
| 1024x1024 (all)      | 22 us   | 34 us   | 136 us            | 147 us        |
| 1024x1024 (0)        | 30 us   | 28 us   | 160 us            | 160 us        |
| 1024x1024 (1)        | 25 us   | 25 us   | 171 us            | 146 us        |
| 1024x10x1024 (all)   | 542 us  | 550 us  | 4.14 ms           | 3.11 ms       |
| 1024x10x1024 (0)     | 658 us  | 690 us  | 6.80 ms           | 5.93 ms       |
| 1024x10x1024 (1)     | 761 us  | 757 us  | 3.34 ms           | 3.52 ms       |
| 1024x10x1024 (2)     | 538 us  | 545 us  | 3.73 ms           | 3.04 ms       |
| 1024x1024x1024 (all) | 72 ms   | 71 ms   | 364 ms            | 357 ms        |
| 1024x1024x1024 (0)   | 94 ms   | 90 ms   | 935 ms            | 927 ms        |
| 1024x1024x1024 (1)   | 80 ms   | 86 ms   | 881 ms            | 688 ms        |
| 1024x1024x1024 (2)   | 71 ms   | 71 ms   | 456 ms            | 354 ms        |

CUDA

| Reduction (dim)      | M40 base | M40 PR  | P100 base | P100 PR   |
|----------------------|----------|---------|-----------|-----------|
| 1024x10x1024 (all)   | 238 us   | 182 us  | 136 us    | 97 us     |
| 1024x10x1024 (0)     | 166 us   | 179 us  | 105 us    | 84 us     |
| 1024x10x1024 (1)     | 181 us   | 182 us  | 89 us     | 91 us     |
| 1024x10x1024 (2)     | 180 us   | 168 us  | 88 us     | 79 us     |
| 1024x1024x1024 (all) | 17.5 ms  | 16.4 ms | 8.23 ms   | 7.48 ms   |
| 1024x1024x1024 (0)   | 27.2 ms  | 28.6 ms | 7.63 ms   | 7.38 ms   |
| 1024x1024x1024 (1)   | 16.5 ms  | 16.3 ms | 7.66 ms   | 7.40 ms   |
| 1024x1024x1024 (2)   | 17.8 ms  | 16.4 ms | 8.37 ms   | 7.31 ms   |

Timings were generated with this script:
https://gist.github.com/colesbury/d3238b266d8a9872fe6f68f77619b379

### CODE REVIEW

**Comment by t-vi** [comment]
Awesome stuff! I can't wait to switch einsum and tensordot to this.

File: aten/src/ATen/native/cpu/Reduce.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// reduction that is contiguous over the input in dim 1
+template <typename traits>
+static inline bool is_outer_reduction(const int64_t* strides) {
+  return strides[0] == 0 &&
+         strides[2] == sizeof(typename traits::result_type) &&
+         strides[3] == sizeof(typename traits::arg2_t);
+}
+
+template <typename func_t, typename vec_func_t>
+void binary_kernel_reduce_vec(TensorIterator& iter, func_t op, vec_func_t vop, int ident=0) {
```
Comment (fmassa) [inline aten/src/ATen/native/cpu/Reduce.h]: question: if we want to apply `TensorIterator` to `min`/`max` reduction, shouldn't we use `ident=-inf / +inf`?
In this case, would it be better to make `ident` of float type to support inf?

Comment (colesbury) [reply aten/src/ATen/native/cpu/Reduce.h]: Yeah, that makes sense

File: aten/src/THC/THCDeviceUtils.cuh [hunk 2]
```diff
@@ -107,6 +107,15 @@ __device__ __forceinline__ T WARP_SHFL_DOWN(T value, unsigned int delta, int wid
     return __shfl_down(value, delta, width);
 #endif
 }
+template <>
+__device__ __forceinline__ at::Half WARP_SHFL_DOWN(at::Half value, unsigned int delta, int width, unsigned int mask)
```
Comment (ngimel) [inline aten/src/THC/THCDeviceUtils.cuh]: It looks like WARP_SHFL_DOWN should never be called with Half argument, so why is this necessary?

File: aten/src/ATen/native/TensorIterator.cpp [hunk 3]
```diff
@@ -1,53 +1,81 @@
...
 
-  // initialize perm with 0, 1, 2, ...
-  perm_.resize(ndim());
-  std::iota(std::begin(perm_), std::end(perm_), 0);
-
-  std::sort(std::begin(perm_), std::end(perm_), [&](size_t i1, size_t i2) {
-    return sum_of_strides[i1] < sum_of_strides[i2];
-  });
+void TensorIterator::reorder_dimensions() {
+  // Sort the dimensions based on strides in ascending order with reduced dims
```
Comment (fmassa) [inline aten/src/ATen/native/TensorIterator.cpp]: It looks like you changed the strategy for permuting the dimensions. Does it affect runtime for pointwise operations?

What if the first argument is the single one that has strides different than the (possibly many) other arguments?

**Comment by fmassa** [comment]
ping @colesbury

File: aten/src/THC/THCDeviceUtils.cuh (hunk 2 above)
Comment (colesbury) [reply aten/src/THC/THCDeviceUtils.cuh]: It's called here in Reduce.cuh:
https://github.com/pytorch/pytorch/pull/11908/files#diff-662693ef7b7f32fa32d7179b6614fc16R296

(The Reduce.cuh file might be collapsed by default in GitHub's "Files changed" view)

File: aten/src/ATen/native/TensorIterator.cpp (hunk 3 above)
Comment (colesbury) [reply aten/src/ATen/native/TensorIterator.cpp]: Previously we sorted based on the sum-of-strides. I changed it to only swap dimensions that make the ordering no-worse across all arguments (and better for at least one argument), with stride 0 not counting and reduced dimensions moved to the front. This more closely matches the stride sorting behavior of THC and NumPy, with the exception of the reduced dimensions behavior.

I don't expect this to affect the runtime for pointwise operations. Most orderings remain the same, and contiguous dims get moved to the front, which is the most important aspect.

I'm not sure this is the final strategy for ordering strides, but this logic makes it easier to support reductions.

File: aten/src/THC/THCDeviceUtils.cuh (hunk 2 above)
Comment (ngimel) [reply aten/src/THC/THCDeviceUtils.cuh]: But `warp_reduce` is called on the argument of type `arg_t` and `arg_t` is https://github.com/pytorch/pytorch/pull/11908/files#diff-662693ef7b7f32fa32d7179b6614fc16R199 (should be no narrower than fp32). I'm (we're) very suspect of any operations directly on halfs, it's precision/overflow/underflow risk. With current reduction ops we made sure that it never happens (all data is converted on read), and it would be great if it did not creep back in.

Comment (colesbury) [reply aten/src/THC/THCDeviceUtils.cuh]: I'll change it to use float32 for intermediate values, but I don't really see how that's going to help. The output is still float16 so you still have overflow issues. You can see this in master with:

```
x = torch.ones(65520, device='cuda', dtype=torch.float16)
x.sum()
```

```
tensor(inf, device='cuda:0', dtype=torch.float16)
```

Comment (colesbury) [reply aten/src/THC/THCDeviceUtils.cuh]: Nevermind. You're right, it already uses float32 for float16 inputs....

File: aten/src/ATen/native/cpu/Reduce.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include <ATen/native/cpu/Loops.h>
+#include <ATen/Parallel.h>
+
+#include <sstream>
+
+namespace at { namespace native { namespace {
```
Comment (gchanan) [inline aten/src/ATen/native/cpu/Reduce.h]: did you want an unnamed namespace in a header file?  (https://wiki.sei.cmu.edu/confluence/display/cplusplus/DCL59-CPP.+Do+not+define+an+unnamed+namespace+in+a+header+file).

File: aten/src/ATen/native/ReduceOps.cpp [hunk 5]
```diff
@@ -116,46 +190,70 @@ Tensor mean(const Tensor &self) {
   return at::native::mean(self, nullopt);
 }
 
-static inline Tensor sum(const Tensor &self, optional<ScalarType> dtype) {
-  return at::_sum(integer_upcast(self, dtype));
+static ScalarType get_dtype(Tensor& result, const Tensor& self, optional<ScalarType> dtype,
+                            bool promote_integers=false) {
+  if (dtype.has_value()) {
+    return dtype.value();
```
Comment (gchanan) [inline aten/src/ATen/native/ReduceOps.cpp]: this is a change of behavior, right?  Previously we checked that if dtype.has_value then dtype.value() == result.type().scalarType().  Why the change?

**Review by gchanan — COMMENTED** [review]
This looks great.   I didn't look through all the math; just have a few questions about functionality.
